### DISCUSSION TITLE
Using cNNP + D3 for geometry optimization in CP2K

### DISCUSSION THREAD

**Discussion by haphysix** [MarsalekGroup/aml#21] [Q&A]
An example for running CP2K-MD with METHOD NNP is included in examples/from-trajectroy-to-model/04-C-NNP-MD.
Would you be able to include a geometry optimization input example with the following requirements?

- MOTION is geometry optimization
- System is periodic
- METHOD is NNP
- D3 dispersion correction is included in energy and force calculations.

Thanks very much!

**Comment by cschran** [comment]
Hi haphysix,

I break down this question into two answers, first for the geometry optimization:

Geometry optimizations work via the [RUN_TYPE](https://manual.cp2k.org/trunk/CP2K_INPUT/GLOBAL.html#list_RUN_TYPE)  `GEO_OPT` 
```
...
RUN_TYPE GEO_OPT
...
```
Replace the MD section for suitable input e.g. from here: https://www.cp2k.org/howto:geometry_optimisation

```
&GEO_OPT
    TYPE MINIMIZATION
    MAX_DR    1.0E-03
    MAX_FORCE 1.0E-03
    RMS_DR    1.0E-03
    RMS_FORCE 1.0E-03
    MAX_ITER 200
    OPTIMIZER CG
    &CG
      MAX_STEEP_STEPS  0
      RESTART_LIMIT 9.0E-01
    &END CG
&END GEO_OPT
```

**Comment by cschran** [comment]
For the addition of different types of interactions, cp2k allows to add various `FORCE_EVAL` sections.
```
&MULTIPLE_FORCE_EVALS
  FORCE_EVAL_ORDER  2 3
  MULTIPLE_SUBSYS  F
&END MULTIPLE_FORCE_EVALS
```
Use the `MIXED` section to define the mixing rules, e.g. simple addition:
```
&FORCE_EVAL
  METHOD  MIXED
  &MIXED
    MIXING_TYPE  GENMIX
    GROUP_PARTITION  1023 1
    &GENERIC
      MIXING_FUNCTION a+b
      VARIABLES a b
    &END GENERIC
  &END MIXED
...
&END FORCE_EVAL
```
and define afterwards what `FORCE_EVAL` you want to mix.
```
&FORCE_EVAL
  METHOD  NNP
...
&END FORCE_EVAL
```
```
&FORCE_EVAL
  METHOD  QS
...
&END FORCE_EVAL
```

**Comment by haphysix** [comment]
Thanks for your reply, Christoph.
At the moment, I am compiling CP2K. Once I have CP2K ready, I'll give it a try.

**Reply by haphysix** [reply]
Thanks for your message, Christoph. The suggested CP2K inputs work just fine. After playing with CP2K inputs for a few days, I know that D3 can only be included in the DFT/XC_FUNCTIONAL block, which means that some unwanted energy and force contributions will be added to NNP values. This will disrupt the geometry optimization process. Could FORCE_EVAL/NNP be configured to include the DFTD3 correction?

**Reply by cschran** [reply]
This is actually more a quickstep CP2K related question and best asked in the CP2K google group, where it might even have been answered.

However, you can try adding a `XC_FUNCTIONAL` dummy section via
```
&XC_FUNCTIONAL NONE
&END XC_FUNCTIONAL
```
